### PR TITLE
qa: test new client with old cluster

### PR DESCRIPTION
**Pull request by dparmar18**
Fixes: https://tracker.ceph.com/issues/53573
Signed-off-by: Dhairya Parmar <dparmar@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
If we're testing a new client with an old cluster, why are we moving to install Pacific (which we're still doing point releases and backports on, so might miss actual protocol issues with) instead of Nautilus (which is a fixed quantity)?

**Comment by dparmar18** [comment]
> If we're testing a new client with an old cluster, why are we moving to install Pacific (which we're still doing point releases and backports on, so might miss actual protocol issues with) instead of Nautilus (which is a fixed quantity)?

I don't precisely remember but I had some discussion with Venky on IRC or 1x1 where we decided to go with Quincy clients with pacific cluster. If pacific can hide such things, I've move back to nautilus but things aren't good with pacific cluster too, almost all the tests are stuck since hours in teuthology http://pulpito.front.sepia.ceph.com/dparmar-2022-10-10_13:22:59-fs:upgrade-main-distro-default-smithi/

**Review by rishabh-d-dave — COMMENTED** [review]
Took a quick look at the failure - http://qa-proxy.ceph.com/teuthology/dparmar-2022-10-11_06:06:11-fs:upgrade-main-distro-default-smithi/7062982/teuthology.log

This what I spot in logs -

```
2022-10-11T06:27:04.579 DEBUG:teuthology.orchestra.run:got remote process result: 1
2022-10-11T06:27:04.580 INFO:teuthology.orchestra.run.smithi170.stderr:chmod: changing permissions of '/home/ubuntu/cephtest/mnt.1': Operation not permitted
```
Above error is from this command - 

```
2022-10-11T06:27:04.521 DEBUG:teuthology.orchestra.run.smithi170:> chmod 0000 /home/ubuntu/cephtest/mnt.1
```


This is where it lies in our codebase - https://github.com/ceph/ceph/blob/main/qa/tasks/cephfs/mount.py#L241. Passing `sudo=True` should do the trick.

Strange that we never hit this issue before.

File: qa/suites/fs/upgrade/upgraded_client/from_nautilus/tasks/2-client-sanity.yaml [hunk 1]
```diff
@@ -1,4 +0,0 @@
-tasks:
-  - cephfs_test_runner:
-      modules:
-        - tasks.cephfs.test_newops
```
Comment (lxbsz) [inline qa/suites/fs/upgrade/upgraded_client/from_nautilus/tasks/2-client-sanity.yaml:4]: Why are you deleting this test ? This is dedicated to test the newops based the nautilus cluster. You should add new file to test the other versions.

File: qa/tasks/cephfs/test_new_client_old_cluster.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+log = logging.getLogger(__name__)
+
+
+class TestNewClientOldCluster(CephFSTestCase):
+    def test_pacific_client_mount(self):
+        if isinstance(self.mount_a, FuseMount):
+            log.info('client is fuse mounted')
+        elif isinstance(self.mount_a, KernelMount):
+            log.info('client is kernel mounted')
```
[the remark this answers is not in the document]
Comment (dparmar18) [reply qa/tasks/cephfs/test_new_client_old_cluster.py:14]: this will be removed soon after im done with the tests

**Comment by dparmar18** [comment]
Okay, so definitely there is some issue with `ceph-fuse`, 

1) http://pulpito.front.sepia.ceph.com/dparmar-2022-10-11_17:01:35-fs:upgrade-main-distro-default-smithi/
2) http://pulpito.front.sepia.ceph.com/dparmar-2022-10-11_17:48:59-fs:upgrade-main-distro-default-smithi/


yamls used in [1]:
```
tasks:
- ceph-fuse: [client.0]
- workunit:
    clients:
       client.0:
         - suites/iozone.sh
```
AND
```
tasks:
- ceph-fuse: [client.0]
- workunit:
    clients:
       client.0:
         - suites/blogbench.sh
```
[1] never completed, had to kill the run. Jobs run into timeout issues and this happens at `mkdir -p -v /home/ubuntu/cephtest/mnt.0`

When I looked into the machine, I cannot list `cephtest` dir, I cannot do `df -hT`. `ceph osd blocklist ls` shows one entry, I suspect it's the client only. Doing `umout -fl home/ubuntu/cephtest/mnt.0` and clearing the blocklist can un-hang the job but ultimately the test would fail.

-----------------

[2] which is all green uses the exact same yaml just `- ceph-fuse: [client.0]` removed, those yamls are - https://github.com/ceph/ceph/pull/48280/commits/7c0d7a28de9f7f0f366a940ccd0af417d6181cd2#diff-a609885c2c09cab572df7a707110532bb78b235691d3c51a28db1128d5f3ae6a and https://github.com/ceph/ceph/pull/48280/commits/7c0d7a28de9f7f0f366a940ccd0af417d6181cd2#diff-f9147c01c698bb81576733e51a936f4ce490271b768450147043808b7baa78b1

and both the jobs succeed without even a single hiccup.

But I can also see this in logs:
```
2022-10-11T18:06:16.269 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi088.stderr:ceph-fuse[39786]: starting ceph client
```
```
2022-10-11T18:07:44.542 INFO:tasks.cephfs.fuse_mount.ceph-fuse.0.smithi057.stderr:ceph-fuse[39794]: starting ceph client
```
If these tests are also working with `ceph-fuse` and PASS, question is why and how there is not hang issue here as I can see it does create dir at mount point successfully:
```
2022-10-11T18:06:15.846 DEBUG:teuthology.orchestra.run.smithi088:> mkdir -p -v /home/ubuntu/cephtest/mnt.0
2022-10-11T18:06:15.867 INFO:teuthology.orchestra.run.smithi088.stdout:mkdir: created directory '/home/ubuntu/cephtest/mnt.0'
```
```
2022-10-11T18:07:44.117 DEBUG:teuthology.orchestra.run.smithi057:> mkdir -p -v /home/ubuntu/cephtest/mnt.0
2022-10-11T18:07:44.140 INFO:teuthology.orchestra.run.smithi057.stdout:mkdir: created directory '/home/ubuntu/cephtest/mnt.0'
```

----------------------------
I've tested this multiple times with different cluster setups like: having client on the same node as other daemons and client on different node and both yield similar results i.e. test hangs at doing `mkdir -p -v` at mountpoint. Another failure that occurred intermittently - (`Command failed on smithi203 with status 1: 'chmod 0000 /home/ubuntu/cephtest/mnt.1'`) which @rishabh-d-dave helped me fix where I had to include `omit_sudo=False` in `create_mntpt()` in `mount.py` .

**Comment by lxbsz** [comment]
> Okay, so definitely there is some issue with `ceph-fuse`,
> 
> 1. http://pulpito.front.sepia.ceph.com/dparmar-2022-10-11_17:01:35-fs:upgrade-main-distro-default-smithi/

It seems you are doing the `ceph-fuse` mount twice:

```
  - ceph-fuse:
      client.0: null
  - print: '**** done remount client'
  - ceph-fuse:
    - client.0
```
I am not sure this will work for the netns, I never test this.

Thanks!

File: qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_dbench.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+tasks:
+- ceph-fuse: [client.0]
```
Comment (lxbsz) [inline qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_dbench.yaml:2]: In `1-client-upgrade.yaml` has already include this:

```
  1 tasks:
  2 - install.upgrade:                                                                                                                                 
  3     client.0:
  4 - print: "**** done install.upgrade on client.0"
  5 - ceph-fuse:
  6     client.0:
  7 - print: "**** done remount client"
```
So here you are mounting the `ceph-fuse: [client.0]` twice.

File: qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_kernel_untar_build.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+tasks:
+- ceph-fuse: [client.0]
```
Comment (lxbsz) [inline qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_kernel_untar_build.yaml:2]: Same here.

**Comment by lxbsz** [comment]
As we discussed please just append the `iozone, dbench, kernel untar and blogbench` test after `newops` test for `from_nautilus/`. And add new directory `from_pacific/` for Pacific cluster, etc.

**Comment by dparmar18** [comment]
I've re-structured `upgraded_client` and added tests. Also addressed all the failures and issues that prevented tests from running successfully.

File: qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_kernel_untar_build.yaml (hunk 4 above)
Comment (dparmar18) [reply qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_kernel_untar_build.yaml]: This has been removed, thanks :)

File: qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_dbench.yaml (hunk 3 above)
Comment (dparmar18) [reply qa/suites/fs/upgrade/upgraded_client/from_pacific/tasks/2-workload/cfuse_dbench.yaml]: Yep, removed. Thanks!

**Comment by dparmar18** [comment]
I was experiencing failures like `Command failed on smithi157 with status 1: 'chmod 0000 /home/ubuntu/cephtest/mnt.1'` in teuthology, therefore added https://github.com/ceph/ceph/pull/48280/commits/5105d1b8fc1d712ae1438724e6529d17ea299034 to address it, should I move this to a complete separate PR or it is okay to ship with this PR? Asking because `qa/tasks/cephfs/mount.py` is widely used, it can/might cause similar failures in future too. Any suggestions/thoughts?

**Comment by dparmar18** [comment]
http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/ - Run with the latest patch. Ignore [7063339](http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/7063339) and [7063346](http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/7063346) they are from `fs:upgrade/featureful_client/upgraded_client` and not `fs:upgrade/upgraded_client` (mistake with `--filter` with `teuthology-suite` cmd). 

All tests on pacific passed while nautilus tests failed with 
```
teuthology.exceptions.CommandFailedError: Command failed on smithi017 with status 22: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd blocklist ls'
``` 
which I think @lxbsz is already working on?

File: qa/suites/fs/upgrade/upgraded_client/from_nautilus/tasks/2-client-sanity.yaml (hunk 1 above)
Comment (dparmar18) [reply qa/suites/fs/upgrade/upgraded_client/from_nautilus/tasks/2-client-sanity.yaml]: Undone the removal

**Comment by lxbsz** [comment]
> http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/ - Run with the latest patch. Ignore [7063339](http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/7063339) and [7063346](http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/7063346) they are from `fs:upgrade/featureful_client/upgraded_client` and not `fs:upgrade/upgraded_client` (mistake with `--filter` with `teuthology-suite` cmd).
> 
> All tests on pacific passed while nautilus tests failed with
> 
> ```
> teuthology.exceptions.CommandFailedError: Command failed on smithi017 with status 22: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd blocklist ls'
> ```
> 
> which I think @lxbsz is already working on?

The PR https://github.com/ceph/ceph/pull/48182 already merged!

**Comment by dparmar18** [comment]
> > http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/ - Run with the latest patch. Ignore [7063339](http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/7063339) and [7063346](http://pulpito.front.sepia.ceph.com/dparmar-2022-10-12_10:31:59-fs:upgrade-main-distro-default-smithi/7063346) they are from `fs:upgrade/featureful_client/upgraded_client` and not `fs:upgrade/upgraded_client` (mistake with `--filter` with `teuthology-suite` cmd).
> > All tests on pacific passed while nautilus tests failed with
> > ```
> > teuthology.exceptions.CommandFailedError: Command failed on smithi017 with status 22: 'sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd blocklist ls'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which I think @lxbsz is already working on?
> 
> The PR #48182 already merged!

Thanks @lxbsz! It should not be an issue then. As discussed this is still pending backports right? we need this to land in nautilus branch else it's not gonna pass in anyway.

**Comment by dparmar18** [comment]
removed commit addressing sudo/su issue, created separate PR https://github.com/ceph/ceph/pull/48476

File: qa/suites/fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/upgraded_client_kernel_untar.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      - elfutils-libelf-devel
+      - openssl-devel
+      - NetworkManager
+      - iproute
+      - util-linux
+- workunit:
+    clients:
+      all:
+        - kernel_untar_build.sh
+
```
Comment (lxbsz) [inline qa/suites/fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/upgraded_client_kernel_untar.yaml:15]: Maybe you could add this file to `.qa/suites/fs/workload/tasks/5-workunit/suites/` and then link it here just likes the `iozone.yaml`?

The others looks nice to me.

**Comment by dparmar18** [comment]
all test pass after rebase (with PR https://github.com/ceph/ceph/pull/48182 in my branch). 

http://pulpito.front.sepia.ceph.com/dparmar-2022-10-15_11:27:50-fs:upgrade-main-distro-default-smithi/

File: qa/suites/fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/upgraded_client_kernel_untar.yaml (hunk 5 above)
Comment (dparmar18) [reply qa/suites/fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/upgraded_client_kernel_untar.yaml:15]: There already exists `kernel_untar_build.yaml` in `.qa/suites/fs/workload/tasks/5-workunit`, but it doesn't install the packages that this test needs, it plain simple `kernel_untar_build` test. So now if i move this to `/tasks/5-workunit`, i hope that is fine as there could be suite(s) where the preparation yamls take care of the installation of these packages and this would then unnecessarily try install them again.

Comment (lxbsz) [reply qa/suites/fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/upgraded_client_kernel_untar.yaml:15]: > There already exists `kernel_untar_build.yaml` in `.qa/suites/fs/workload/tasks/5-workunit`, but it doesn't install the packages that this test needs, it plain simple `kernel_untar_build` test. So now if i move this to `/tasks/5-workunit`, i hope that is fine as there could be suite(s) where the preparation yamls take care of the installation of these packages and this would then unnecessarily try install them again.

You should fix it anyway in `.qa/suites/fs/workload/tasks/5-workunit`. Just make it to work in all the cases, like the one of this PR hit. IMO it's no harm to always install the depending packages.

Comment (dparmar18) [reply qa/suites/fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/upgraded_client_kernel_untar.yaml:15]: Alright, I'll make the change

Comment (dparmar18) [reply qa/suites/fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/upgraded_client_kernel_untar.yaml]: Done, thanks.

**Review by lxbsz — APPROVED** [review]
Maybe we should just link the files by keeping the old names as they are as we do in most of other test cases, for example just do:

```
.qa/suites/fs/workload/tasks/5-workunit/iozone.yaml --> fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/iozone.yaml
```
But it's trivial.

**Comment by dparmar18** [comment]
> 

I'll rename them

**Comment by dparmar18** [comment]
> Maybe we should just link the files by keeping the old names as they are as we do in most of other test cases, for example just do:
> 
> ```
> .qa/suites/fs/workload/tasks/5-workunit/iozone.yaml --> fs/upgrade/upgraded_client/tasks/2-workload/stress_tests/iozone.yaml
> ```
> 
> But it's trivial.

done

**Review by lxbsz — APPROVED** [review]
LGTM. Thanks!

**Review by rishabh-d-dave — APPROVED** [review]
Looks good.

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#09-Mar-2023

File: qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml [hunk 6]
```diff
@@ -4,6 +4,15 @@ overrides:
       mds:
         - "mds.dir_split"
 tasks:
+- install:
```
Comment (vshankar) [inline qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml:7]: @dparmar18 Is this change related?

Comment (dparmar18) [reply qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml:7]: yes these packages are very important, else the kernel untar tests would fail

Comment (vshankar) [reply qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml:7]: Hmmm.. why aren't those failing now. Needs explanation in commit.

Comment (dparmar18) [reply qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml:7]: because the suite that has untar tests also install these packages e.g. `qa/suites/fs/mixed-clients`, symlink to https://github.com/ceph/ceph/blob/main/qa/cephfs/begin/0-install.yaml can be found in it

Comment (dparmar18) [reply qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml:7]: ok ill update it

Comment (dparmar18) [reply qa/suites/fs/workload/tasks/5-workunit/kernel_untar_build.yaml:7]: wait, instead of making change to a generic yaml file, i can add these packages to `0-nautilus.yaml` and `0-pacific.yaml`

**Comment by dparmar18** [comment]
changes made: https://github.com/ceph/ceph/compare/8df3225a2dbbe7ad482f8847891dff1ad352d64d..179e4bcae9bc7b6b0c049120ed5795c143180e0a

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#09-Mar-2023
